refactor: tsconfck for esbuild compilerOptions tsconfig files

dominikg opened the pull request:
<!-- Thank you for contributing! -->

### Description

* replace custom tsconfig parsing in esbuild plugin with `tsconfck` package. (Disclaimer: i wrote tsconfck)
* add error handling to continue when parsing tsonfig fails

why:

* greatly reduces code in esbuild.ts
* adds support for solution-style tsconfig files
* uses more effective caching to reduce repeated parse for multiple files sharing a tsconfig
* has a testsuite that validates it works like typescripts native functions

### Additional context

[tsconfck](https://github.com/dominikg/tsconfck) "extends" and "parse" implementation is based on how vite currently works, with additions for rebasing relative paths and parsing solution-style tsconfigs.

---

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [ ] Bug fix
- [ ] New Feature
- [ ] Documentation update
- [x] Other

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md#pull-request-guidelines) and follow the [Commit Convention](https://github.com/vitejs/vite/blob/main/.github/commit-convention.md).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [ ] Ideally, include relevant tests that fail without this PR but pass with it.

dominikg committed:
wip: use tsconfck to load tsconfig

dominikg committed:
wip: add error handling and testcase for tsconfig parsing

dominikg commented:
The implementation aggressively caches parsed results, just like the previous vite one. 

I wonder if the cache has to be cleared if a tsconfig file changes. Didn't see a quick way to add watchers here, so that should be discussed more before merging this maye.

cc @sodatea

dominikg commented:
Oh, tsconfck isn't released as 1.0.0 right now, but it is complete i think. Pending some Feedback for the general usability of it's api i can release 1.0.0 before merging this.

Shinigami92 commented:
I have two thoughts/questions:

1. I think we can trust you (you are a known community member) but should we wait until the `1.0.0` was released?
2. What does the rest of the vite team think of this new dependency?

sodatea commented on .prettierignore:
```diff
@@ -4,4 +4,5 @@ packages/plugin-vue/dist/
-yarn.lock
\ No newline at end of file
+yarn.lock
+packages/playground/tsconfig-json/has-error/tsconfig.json
```
I think it's better to have a `// prettier-ignore` comment in the `tsconfig.json`.

dominikg replied:
unfortunately that does not work. The parser error trips up prettier and it never reads the ignore comment.

sodatea commented on packages/vite/src/node/plugins/esbuild.ts:
```diff
@@ -200,192 +197,22 @@ function prettifyMessage(m: Message, code: string): string {
+  return parse(filename, {
+    cache: tsconfigCache,
+    resolveWithEmptyIfConfigNotFound: true
+  }).then(
```
Now that it's inside an `async` function, plain `try { await ... } catch` seems more intuitive.

dominikg replied:
uses try catch now

sodatea commented on packages/vite/src/node/plugins/esbuild.ts:
```diff
@@ -200,192 +197,22 @@ function prettifyMessage(m: Message, code: string): string {
+    (err) => {
+      debug(`error while parsing tsconfig for ${filename}, ${err}`)
+      tsconfigCache.set(filename, {
+        filename: 'tsconfig_parse_error',
```
I find it a little confusing to have 2 `filename`s with completely different meanings. How about renaming the first `filename` to something like` srcFilePath`? Or the second one to `configPath`?

dominikg replied:
this is something i could improve in tsconfck, thanks for pointing out a source of confusion

dominikg replied:
it is called `tsconfigFile` in the result object now.

sodatea commented on packages/vite/src/node/plugins/esbuild.ts:
```diff
@@ -200,192 +197,22 @@ function prettifyMessage(m: Message, code: string): string {
+  }).then(
+    (result) => result.tsconfig,
+    (err) => {
+      debug(`error while parsing tsconfig for ${filename}, ${err}`)
```
Should be a hard error. Most people never see the `debug` log.

sodatea replied:
`throw new Error` or `logger.error`. I'm not sure which is better though.

dominikg replied:
Yeah, i was a bit wary about adding error logs. some of the errors (no config found) were silently swallowed before so i didn't want to introduce new ones. But you're right that they should be more visible.

Throw is fine on build, on dev maybe log would be better, unless the server recovers from throw after user fixed tsconfig

dominikg replied:
i have added throwing errors and config file watching. fixing config errors while devserver is running reloads everything.

sodatea commented:
Thanks for the help!
I think it's super cool to finally have a full-fledged tsconfig parser, and with `solutions` support!

The above comments are just my nitpicky suggestions - the overall functionality looks good to me.

sodatea commented:
> I wonder if the cache has to be cleared if a tsconfig file changes. Didn't see a quick way to add watchers here, so that should be discussed more before merging this maye.

```js
return {
  name: 'vite:esbuild',
  configureServer(server) {
    server.watcher
	// ...
  },
}
```

Like this?

patak-js commented:
Maybe we could try to merge this one as part of the 2.6 beta 👍

dominikg committed:
wip: throw errors when parsing tsconfig, watch for tsconfig changes and reload

dominikg commented:
added config watching and throwing errors. 
Also made a separate test playground to verify error overlay is shown.
After edit, server reloads and code is compiled. :partying_face:

dominikg commented:
example: 
![Screenshot_2021-09-14_16-14-27](https://user-images.githubusercontent.com/611613/133274218-a3b12e80-fdcf-4855-9bc0-2dca3d46af44.png)

dominikg committed:
fix: update assertion to be os agnostic.

dominikg committed:
fix: don't run page tests on failed build

dominikg committed:
fix: allow tests to see and remove error thrown in beforeAll

Shinigami92 commented on packages/vite/package.json:
```diff
@@ -116,10 +116,8 @@
-    "strip-json-comments": "^3.1.1",
     "terser": "^5.7.1",
-    "tsconfig": "^7.0.0",
+    "tsconfck": "^1.0.0-10",
```
Would it be okay for you to release a _"stable"_ `1.0.0` before merging this?

dominikg commented:
yes. I already updated the api after @sodatea s comment about filename being ambiguous, and added tsconfigFile to the error output. if there is nothing else you're worried about I'li spin 1.0

dominikg committed:
chore: update to tsconfck 1.0.0 and remove outdated comment from test

dominikg commented:
and there's tsconfck 1.0. Updated the license to include inlined code by Sindre Sorhus so it's picked up correctly by vite's license generator.

If it is fine to do a nuclear reload on every tsconfig change this is ready to go. I don't think it's possible to try and selectively reload files that may have changed, so the only other option would be to accept stale entries, but that would leave us with a weird state where some modules were transpiled with different settings than others.

patak-cat committed:
Merge branch 'main' into refactor/tsconfck
